proposals/0055-stale-series-compaction.md

```diff
@@ -0,0 +1,108 @@
+
+We can keep a running counter that tracks how many series are stale at the moment. Incremented or decremented based on the incoming sample and the last sample of the series.
+
+### Compacting Stale Series
+
+We will have a single thresholds to trigger stale series compaction, `R`, which is the ratio of stale series count to total series count. It will be configurable and default to 0 (meaning stale series compaction is disabled).
+
+As soon as the ratio of stale series to total series reaches `R`, we trigger the stale series compaction that simply flushes these stale series into a block and removes it from the Head block (can be more than one block if the series crosses the block boundary). We skip WAL truncation and m-map files truncation at this stage and let the usual compaction cycle handle it.
+
+While removing the stale series from the head block, we add tombstones only in the WAL for these stale series with deleted time range as `[MinInt64, MaxInt64]`. WAL replay will simply drop these series from the head as soon as it encounters this record in the WAL. This way we don't spike up the memory during WAL replay.
+
+Since these are stale series, there won’t be any races when compacting it in most cases. We will still lock the series and take required measures so that we don’t cause race with an incoming sample for any stale series.
+
+Implementation detail: if the usual head compaction is about to happen very soon, we should skip the stale series compaction and simply wait for the usual head compaction.
+
+### Experimental Analysis and Trade-offs
+
+Running Prombench on https://github.com/prometheus/prometheus/pull/16929 and running a version of that at Reddit uncovered a few learnings
+
+* When stale series compaction runs, a large number of instant queries can increase both the CPU and memory until the next regular compaction. This is because instant queries will now have to merge data from the memory and the stale series block from disk.
+* When the `R` is set close to the peak of possible stale series ratio, it might actually use more memory than it would have at the peak stale series ratio. This was observed with `R=0.2` internally at Reddit where at peak the Prometheus would have held 30% stale series. Since there are a lot of rules running, holding 30% stale series took less memory than running stale series compaction at `R=0.2`.
+
+**Conclusions**
```

Comment:
The conclusion makes sense to me. I think the flag sort of needs an "advanced" disclaimer to it because this behavior is not straightforward to the user.